nucliadb/nucliadb/ingest/orm/node.py

```diff
@@ -351,3 +444,62 @@ async def chitchat_update_node(members: List[ClusterMember]) -> None:
             if node is not None:
                 logger.info(f"{key}/{node.type} remove {node.address}")
                 await Node.destroy(key)
+
+
+def update_shards_with_shadow_replica(
+    shards: PBShards, replica_id: str, shadow_replica: ShardId, shadow_node: str
+) -> bool:
+    """
+    Returns whether it found the replica to update in the shards proto message
+    """
+    replica = get_replica(shards, replica_id)
+    if replica is None:
+        return False
+    replica.has_shadow = True
+    replica.shadow_replica.node = shadow_node
+    replica.shadow_replica.shard.CopyFrom(shadow_replica)
+    return True
+
+
+def cleanup_shadow_replica_from_shards(shards: PBShards, replica_id: str) -> bool:
+    """
+    Returns whether it found the replica to cleanup in the shards proto message
+    """
+    replica = get_replica(shards, replica_id)
+    if replica is None:
+        return False
+    replica.has_shadow = False
```

Comment:
Has Shadow is duplicated of hasfield on replica?

Reply:
🤦🏼 totally. Still getting familiar with proto python api... fixing